### PR TITLE
resource/aws_mq_broker: Add RabbitMQ engine type

### LINKED ISSUE #16027: Support for RabbitMQ engine on aws_mq_broker
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add support for RabbitMQ as an engine on the aws_mq_broker resource

### New or Affected Resource(s)

* aws_mq_broker

### References

https://aws.amazon.com/blogs/aws/amazon-mq-update-new-rabbitmq-message-broker-service/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+AmazonWebServicesBlog+%28Amazon+Web+Services+Blog%29

### LINKED ISSUE #16030: Support for RabbitMQ in AmazonMQ
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AmazonMQ now [support RabbitMQ](https://aws.amazon.com/about-aws/whats-new/2020/11/announcing-amazon-mq-rabbitmq/) as an engine. This is the first time AWS adds another type of engine to AMQ service. Looking at the Go package - [the changes are minimal](https://github.com/aws/aws-sdk-go/commit/d0f43158ad3bcbe89ab04b7fdef545cf7a591154#diff-66b4ca02d46933f68848c054e2b618e90b40be2cbb05e2c1d93aa91b83ce7d93).

### New or Affected Resource(s)

* aws_mq_broker

### Potential Terraform Configuration

```hcl
resource "aws_mq_broker" "example" {
  broker_name        = "example"

  engine_type        = "RabbitMQ"
  engine_version     = "3.8.6"
  host_instance_type = "mq.t2.micro"
  security_groups    = [aws_security_group.test.id]

  user {
    username = "ExampleUser"
    password = "MindTheGap"
  }
}
```

### References

* https://aws.amazon.com/about-aws/whats-new/2020/11/announcing-amazon-mq-rabbitmq/
* https://github.com/aws/aws-sdk-go/commit/d0f43158ad3bcbe89ab04b7fdef545cf7a591154#diff-66b4ca02d46933f68848c054e2b618e90b40be2cbb05e2c1d93aa91b83ce7d93R2138

### LINKED ISSUE #16047: Terraform 0.13 crashes on AWS MQ Import
<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform CLI and Terraform AWS Provider Version
```bash
$ terraform -v
Terraform v0.13.5
```

### Configuration files
```hcl
provider "aws" {
  version = "3.13.0"
  region = var.aws_region
}

resource "aws_mq_broker" "broker" {
}
```

### Affected Resource(s)
* `aws_mq_broker`

### Debug Output

```bash
aws_mq_broker.broker: Refreshing state... [id=broker-staging]                                                                                                                                    [32/1443]
                                                                                                                                                                                                           
Error: rpc error: code = Unavailable desc = transport is closing                                                                                                                                           
                                                                                                                                                                                                           
                                                                                                                                                                                                           
panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                                    
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x42755e3]                                                         
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:                                                                                                                                 
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: goroutine 56 [running]:                                                                                                         
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsMqBrokerRead(0xc000e24f00, 0x530bee0, 0xc0013d9180, 0xa0c0290, 0x604acb0)                                                                                                                                                                                             
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_mq_broker.go:378 +0xbe3                                                                                                                                                                                          
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc000610a50, 0x6a89a00, 0xc000eecb40, 0xc000e24f00, 0x530bee0, 0xc0013d9180, 0x0, 0x0, 0x0)                                                                                                                                                                   
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.1.0/helper/schema/resource.go:280 +0x88                                                                                                                                                                                             
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc000610a50, 0x6a89a00, 0xc000eecb40, 0xc001354310, 0x530bee0, 0xc0013d9180, 0xc001550198, 0x0, 0x0, 0x0)                                                                                                                                    
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.1.0/helper/schema/resource.go:554 +0x1c2                                                                                                                                                                                            
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: github.com/hashicorp/terraform-plugin-sdk/v2/internal/helper/plugin.(*GRPCProviderServer).ReadResource(0xc000d77a40, 0x6a89a00, 0xc000eecb40, 0xc000e00cc0, 0xc000d77a40, 0xc000d77a50, 0x604aca0)                                                                                                                                         
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.1.0/internal/helper/plugin/grpc_provider.go:572 +0x42f                                                                                                                                                                              
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: github.com/hashicorp/terraform-plugin-sdk/v2/internal/tfplugin5._Provider_ReadResource_Handler.func1(0x6a89a00, 0xc000eecb40, 0x5ce7d20, 0xc000e00cc0, 0xc000eecb40, 0x5492960, 0xc000e00c01, 0xc0012103a0)                                                                                                                                
2020-11-05T14:44:05.016Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.1.0/internal/tfplugin5/tfplugin5.pb.go:3276 +0x86                                                                                                                                                                                   
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: github.com/hashicorp/terraform-plugin-sdk/v2/plugin.Serve.func3.1(0x6a89ac0, 0xc000e1f920, 0x5ce7d20, 0xc000e00cc0, 0xc001210380, 0xc0012103a0, 0xc001007ba0, 0x4927f8, 0x5b390e0, 0xc000e1f920)                                                                                                                                           
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.1.0/plugin/serve.go:76 +0x87                                                                                                                                                                                                        
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: github.com/hashicorp/terraform-plugin-sdk/v2/internal/tfplugin5._Provider_ReadResource_Handler(0x5dcb7c0, 0xc000d77a40, 0x6a89ac0, 0xc000e1f920, 0xc000e00c60, 0xc001418ee0, 0x6a89ac0, 0xc000e1f920, 0xc000e3e160, 0x15c)
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.1.0/internal/tfplugin5/tfplugin5.pb.go:3278 +0x14b
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: google.golang.org/grpc.(*Server).processUnaryRPC(0xc000dca700, 0x6aacf40, 0xc001ae4000, 0xc000e0bd00, 0xc000c4e9f0, 0xa084bd0, 0x0, 0x0, 0x0)
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1171 +0x50a
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: google.golang.org/grpc.(*Server).handleStream(0xc000dca700, 0x6aacf40, 0xc001ae4000, 0xc000e0bd00, 0x0)
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1494 +0xccd
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000c461e0, 0xc000dca700, 0x6aacf40, 0xc001ae4000, 0xc000e0bd00)
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:834 +0xa1
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5: created by google.golang.org/grpc.(*Server).serveStreams.func1
2020-11-05T14:44:05.017Z [DEBUG] plugin.terraform-provider-aws_v3.13.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:832 +0x204
2020/11/05 14:44:05 [ERROR] eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2020/11/05 14:44:05 [ERROR] eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2020/11/05 14:44:05 [TRACE] [walkImport] Exiting eval tree: import aws_mq_broker.broker result
2020/11/05 14:44:05 [TRACE] vertex "import aws_mq_broker.broker result": visit complete
2020/11/05 14:44:05 [TRACE] vertex "aws_mq_broker.broker (import id \"broker-staging\")": dynamic subgraph encountered errors
2020/11/05 14:44:05 [TRACE] vertex "aws_mq_broker.broker (import id \"broker-staging\")": visit complete
2020/11/05 14:44:05 [TRACE] vertex "aws_mq_broker.broker (import id \"broker-staging\")": dynamic subgraph encountered errors
2020/11/05 14:44:05 [TRACE] vertex "aws_mq_broker.broker (import id \"broker-staging\")": visit complete
2020/11/05 14:44:05 [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/hashicorp/aws\"] (close)" errored, so skipping
2020/11/05 14:44:05 [TRACE] dag/walk: upstream of "root" errored, so skipping
2020-11-05T14:44:05.020Z [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2020-11-05T14:44:05.020Z [DEBUG] plugin: plugin process exited: path=.terraform/plugins/registry.terraform.io/hashicorp/aws/3.13.0/linux_amd64/terraform-provider-aws_v3.13.0_x5 pid=29581 error="exit status 2"
2020-11-05T14:44:05.067Z [DEBUG] plugin: plugin exited
```

### Panic Output

[crash.log](https://gist.github.com/estebanz01/c06685f1952d4d2b7f6f83b9ef02bca0)

### Expected Behavior
Import AWS MQ broker configuration

### Actual Behavior
Crash

### Steps to Reproduce
Although I'm using `terragrunt`, it also happens if I use terraform instead.

1. `terraform import aws_mq_broker.broker broker-staging`

### Important Factoids
Broker engine RabbitMQ version 3.8.6.

### References
Same error, different components:
* #15239
* #15244
* #10565

### PR DESCRIPTION
**Pull request by yardensachs**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #16030
Closes #16027
Closes #16047

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_mq_broker: Added RabbitMQ engine type
```

Output from acceptance testing:

```% make testacc TESTARGS='-run=TestAccAWSMqBroker_rabbitmq'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSMqBroker_rabbitmq -timeout 120m
=== RUN   TestAccAWSMqBroker_rabbitmq
=== PAUSE TestAccAWSMqBroker_rabbitmq
=== CONT  TestAccAWSMqBroker_rabbitmq
--- PASS: TestAccAWSMqBroker_rabbitmq (514.93s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	518.598s
```

### CODE REVIEW

**Comment by yardensachs** [comment]
@DrFaust92 I was wondering if you might have a few minutes to help me with an issue related to AWS api:

It looks like for RabbitMQ engine type, AWS is not planning on sending the users list in the DescribeBroker API call.

The issue that I am getting is that if the users list does not come back - terraform will assume it needs to add them again, because the mq broker state has that user list empty:
```
make testacc TESTARGS='-run=TestAccAWSMqBroker_rabbitmq'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSMqBroker_rabbitmq -timeout 120m
=== RUN   TestAccAWSMqBroker_rabbitmq
=== PAUSE TestAccAWSMqBroker_rabbitmq
=== CONT  TestAccAWSMqBroker_rabbitmq
    resource_aws_mq_broker_test.go:825: Step 1/2 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
        stdout


        An execution plan has been generated and is shown below.
        Resource actions are indicated with the following symbols:
          ~ update in-place

        Terraform will perform the following actions:

          # aws_mq_broker.test will be updated in-place
          ~ resource "aws_mq_broker" "test" {
                apply_immediately          = false
                arn                        = "arn:aws:mq:us-west-2:810560764000:broker:tf-acc-test-g6wq3:b-5d87a1ba-03da-4522-82de-77c0a7f8017d"
                auto_minor_version_upgrade = false
                broker_name                = "tf-acc-test-g6wq3"
                deployment_mode            = "SINGLE_INSTANCE"
                engine_type                = "RabbitMQ"
                engine_version             = "3.8.6"
                host_instance_type         = "mq.t3.micro"
                id                         = "b-5d87a1ba-03da-4522-82de-77c0a7f8017d"
                instances                  = [
                    {
                        console_url = "https://b-5d87a1ba-03da-4522-82de-77c0a7f8017d.mq.us-west-2.amazonaws.com"
                        endpoints   = [
                            "amqps://b-5d87a1ba-03da-4522-82de-77c0a7f8017d.mq.us-west-2.amazonaws.com:5671",
                        ]
                        ip_address  = ""
                    },
                ]
                publicly_accessible        = false
                security_groups            = [
                    "sg-0f0f5352a4f4b4810",
                ]
                subnet_ids                 = [
                    "subnet-f5064ede",
                ]
                tags                       = {}

                encryption_options {
                    use_aws_owned_key = true
                }

                logs {
                    audit   = false
                    general = false
                }

                maintenance_window_start_time {
                    day_of_week = "TUESDAY"
                    time_of_day = "12:00"
                    time_zone   = "UTC"
                }

              + user {
                  + console_access = false
                  + groups         = []
                  + password       = (sensitive value)
                  + username       = "Test"
                }
            }

        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccAWSMqBroker_rabbitmq (481.57s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	484.127s
FAIL
make: *** [testacc] Error 1
...
```

This only happens with RabbitMQ, It does return on the user list if the engine type is ApacheMQ.

How should I proceed?
Is there a way to ignore that user list? (based on the engine type)
Should username and password be only be effective on broker creation? (based on the engine type)

**Comment by yardensachs** [comment]
@gdavison can you help?

**Comment by yardensachs** [comment]
Ok I figured it out using `DiffSuppressFunc`

**Comment by yardensachs** [comment]
Also tested the whole module:
```
% make testacc TESTARGS='-run=TestAccAWSMqBroker_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSMqBroker_ -timeout 120m
=== RUN   TestAccAWSMqBroker_basic
=== PAUSE TestAccAWSMqBroker_basic
=== RUN   TestAccAWSMqBroker_allFieldsDefaultVpc
=== PAUSE TestAccAWSMqBroker_allFieldsDefaultVpc
=== RUN   TestAccAWSMqBroker_allFieldsCustomVpc
=== PAUSE TestAccAWSMqBroker_allFieldsCustomVpc
=== RUN   TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== RUN   TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== RUN   TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== RUN   TestAccAWSMqBroker_updateUsers
=== PAUSE TestAccAWSMqBroker_updateUsers
=== RUN   TestAccAWSMqBroker_updateTags
=== PAUSE TestAccAWSMqBroker_updateTags
=== RUN   TestAccAWSMqBroker_updateSecurityGroup
=== PAUSE TestAccAWSMqBroker_updateSecurityGroup
=== RUN   TestAccAWSMqBroker_disappears
=== PAUSE TestAccAWSMqBroker_disappears
=== RUN   TestAccAWSMqBroker_rabbitmq
=== PAUSE TestAccAWSMqBroker_rabbitmq
=== CONT  TestAccAWSMqBroker_basic
=== CONT  TestAccAWSMqBroker_updateUsers
=== CONT  TestAccAWSMqBroker_updateTags
=== CONT  TestAccAWSMqBroker_disappears
=== CONT  TestAccAWSMqBroker_rabbitmq
=== CONT  TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== CONT  TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== CONT  TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== CONT  TestAccAWSMqBroker_updateSecurityGroup
=== CONT  TestAccAWSMqBroker_allFieldsCustomVpc
=== CONT  TestAccAWSMqBroker_allFieldsDefaultVpc
2020/11/14 21:42:23 [WARN] Truncating attribute path of 0 diagnostics for TypeSet
--- PASS: TestAccAWSMqBroker_rabbitmq (480.81s)
--- PASS: TestAccAWSMqBroker_disappears (1137.15s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1137.16s)
--- PASS: TestAccAWSMqBroker_basic (1171.67s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1185.68s)
--- PASS: TestAccAWSMqBroker_updateTags (1200.75s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1227.58s)
--- PASS: TestAccAWSMqBroker_updateSecurityGroup (1286.17s)
--- PASS: TestAccAWSMqBroker_updateUsers (1517.43s)
--- PASS: TestAccAWSMqBroker_allFieldsDefaultVpc (1816.32s)
--- PASS: TestAccAWSMqBroker_allFieldsCustomVpc (1826.47s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1830.391s
```

**Review by grigorov — APPROVED** [review]

**Comment by yardensachs** [comment]
@ewbankkit Is this waiting for a bot approval?

**Comment by jonten** [comment]
@yardensachs Just a question. Should'nt `mq.EngineTypeRabbitmq` be added here too? https://github.com/hashicorp/terraform-provider-aws/blob/be8f93c0bc3a1b2d54b3544a7c5d41411ac0f2d4/aws/resource_aws_mq_configuration.go#L59

**Comment by yardensachs** [comment]
> @yardensachs Just a question. Should'nt `mq.EngineTypeRabbitmq` be added here too?
> 
> https://github.com/hashicorp/terraform-provider-aws/blob/be8f93c0bc3a1b2d54b3544a7c5d41411ac0f2d4/aws/resource_aws_mq_configuration.go#L59

The configuration block/resource does not apply to RabbitMQ as far as I'm aware.

Screen Shot 2020-11-19 at 7 00 15 AM

**Comment by yardensachs** [comment]
> @yardensachs Just a question. Should'nt `mq.EngineTypeRabbitmq` be added here too?
> 
> https://github.com/hashicorp/terraform-provider-aws/blob/be8f93c0bc3a1b2d54b3544a7c5d41411ac0f2d4/aws/resource_aws_mq_configuration.go#L59

@jonten The configuration block/resource does not apply to RabbitMQ as far as I'm aware.

Screen Shot 2020-11-19 at 7 00 15 AM

**Comment by jonten** [comment]
> > @yardensachs Just a question. Should'nt `mq.EngineTypeRabbitmq` be added here too?
> > https://github.com/hashicorp/terraform-provider-aws/blob/be8f93c0bc3a1b2d54b3544a7c5d41411ac0f2d4/aws/resource_aws_mq_configuration.go#L59
> 
> @jonten The configuration block/resource does not apply to RabbitMQ as far as I'm aware.
> 
> Screen Shot 2020-11-19 at 7 00 15 AM

Ok, good 👍

**Comment by one70six** [comment]
I tried to use your branch to provision and came across 2 issues:

1. Enabling Cloudwatch logs with `logs { general = true }` results in a plan incuding:
```sh
+ logs {
          + audit   = false
          + general = true
        }
```
which is correct but it errors on the audit log even though you explicitly set `audit = false` or even leave it off:
```sh
aws_mq_broker.tf-rmq-broker: Creating...

Error: BadRequestException: Audit logging is not supported for RabbitMQ brokers.
{
  RespMetadata: {
    StatusCode: 400,
    RequestID: "4a17e8b5-8c0f-4ecd-a881-b058a610f959"
  },
  ErrorAttribute: "logs.audit",
  Message_: "Audit logging is not supported for RabbitMQ brokers."
}
```

2. Setting `publicly_accessible = true` results in:
```sh
aws_mq_broker.tf-rmq-broker: Creating...

Error: BadRequestException: Broker with [publiclyAccessible] set to true does not support specifying [securityGroups]
{
  RespMetadata: {
    StatusCode: 400,
    RequestID: "97343aef-5455-4f2a-94b3-1e49b28e40b8"
  },
  ErrorAttribute: "securityGroups",
  Message_: "Broker with [publiclyAccessible] set to true does not support specifying [securityGroups]"
}
```
and if you remove the security_groups argument:
```sh
Error: Missing required argument

  on main.tf line 6, in resource "aws_mq_broker" "tf-rmq-broker":
   6: resource "aws_mq_broker" "tf-rmq-broker" {

The argument "security_groups" is required, but no definition was found.
```
or if you try setting it to empty:
```sh
aws_mq_broker.tf-rmq-broker: Creating...

Error: BadRequestException: Empty list of security groups is invalid.
{
  RespMetadata: {
    StatusCode: 400,
    RequestID: "a8e6c4fb-05cb-4e06-8bc6-2e6d0243e373"
  },
  ErrorAttribute: "securityGroups",
  Message_: "Empty list of security groups is invalid."
}
```

Other than that it works great!

**Comment by one70six** [comment]
I found that for at least the security_groups issue, the aws-sdk-go that is being called by the hashicorp/aws Terraform module is enforcing that the SecurityGroups is NOT optional. This is most likely intentional for ActiveMQ but breaks in the case of Rabbit? I could be wrong..., bottom line is that there is a conflict between the enforcement on the SDK/API side vs. the Terraform provider side. Bummer...

See: https://github.com/aws/aws-sdk-go/blob/master/service/mq/api.go#L2789

**Comment by alex-bes** [comment]
hi there! thank you very much for the efforts on this! 

is there an ETA on when this will be merged/released? Eager to try it out! :)

**Comment by yardensachs** [comment]
I don't know when this will be merged.
I have other PRs waiting for months, so this might take long as well.

**Comment by yardensachs** [comment]
I don't know when this will be merged.
I have other PRs waiting for months, so this might take long as well.

I see people commented on this.
I am not very good at this yet, and not sure I can resolve the issues @one70six found.

I did this work a long time ago when I had the time to do it.
Now because of the slow response of: commit/review/merge - its hard to keep maintaining this PR and be responsive.

I can't imagine the amount of work the maintainers have, and the speed at which AWS is now releasing changes - I am sure its crazy.

**Comment by sumitg892** [comment]
Hello, @hashicorp/terraform-aws, can we please get it reviewed and merged if possible?

File: website/docs/r/mq_broker.html.markdown
Comment (lukaszkorecki) [inline website/docs/r/mq_broker.html.markdown:70]: It might be helpful to add that RabbitMQ comes with [the management plugin enabled](https://docs.aws.amazon.com/amazon-mq/latest/developer-guide/rabbitmq-basic-elements-plugins.html), so it's also possible to the API to created/update/delete users (and more, `/api/definitions` endpoint is super handy  for that).

We're using the [cloud formation workaround](https://github.com/hashicorp/terraform-provider-aws/issues/16030#issuecomment-778554380) at the moment, and also have a Lambda in place (triggered by Terraform) to set up a couple of exchanges and users.

Comment (DrFaust92) [reply website/docs/r/mq_broker.html.markdown:70]: Unrelated but you can use the rabbit provider to create exchanges and users :D

**Comment by TheGreenSourcerer** [comment]
Any chance to merge it this year? 😄

File: website/docs/r/mq_broker.html.markdown
Comment (lukaszkorecki) [reply website/docs/r/mq_broker.html.markdown:70]: Ah yes, assuming that where the place where you run `terraform apply` has direct access to the RMQ server.

**Comment by brunobraga-hotmart** [comment]
Any merge estimate?

**Comment by YakDriver** [comment]
@yardensachs Thank you for your work on this PR! There are many limitations currently with using RabbitMQ in AWS currently. However, if we provide the capability and document the limits carefully, people can decide for themselves.

I cannot guarantee how soon I'll get to this. However, to facilitate the process, here are the limitations I see in the documentation:

1. For `BrokerInstance`, a RabbitMQ broker instance can't have an associated IP address (`IpAddress`) of the ENI attached to the broker.
2. For `CreateBrokerRequest`, the list of information about broker configuration (`Configuration`) does not apply to RabbitMQ brokers.
3. For `CreateBrokerRequest`, LDAP server metadata (`LdapServerMetadata`) to authenticate and authorize connections to a broker is not supported for RabbitMQ.
4. For `CreateBrokerRequest`, the storage type (`StorageType`) cannot be EFS for RabbitMQ brokers.
5. For `CreateConfiguration`, `CreateConfigurationRequest`, the engine type can be RabbitMQ. Perhaps an error? See #2.
6. For `Logs`, audit logging (`Audit`) does not apply to RabbitMQ, while general logging (`General`) does appear to be applicable to RabbitMQ.
7. For `UpdateBrokerRequest`, the list of information about broker configuration (`Configuration`) does not apply to RabbitMQ brokers. See #2.
8. For `UpdateBrokerRequest `, LDAP server metadata (`LdapServerMetadata`) to authenticate and authorize connections to a broker is not supported for RabbitMQ. See #3.
9. For `User`, a RabbitMQ user cannot have console access (`ConsoleAccess`)
10. For `User`, a RabbitMQ user cannot be associated with a list of groups (`Groups`). (`Username` and `Password` do not list RabbitMQ limitations.)

**Comment by YakDriver** [comment]
@one70six This is super helpful insight!

> I found that for at least the security_groups issue, the aws-sdk-go that is being called by the hashicorp/aws Terraform module is enforcing that the SecurityGroups is NOT optional. 

Indeed, `security_groups` is marked as required althought the AWS documentation does not (now) list them as required. The documentation may be wrong or they may have changed things. We should be able to figure it out though.

> which is correct but it errors on the audit log even though you explicitly set audit = false or even leave it off:

This is easy to fix. Currently we always send `audit` whether it is there or not. It should not.

**Comment by YakDriver** [comment]
@yardensachs On second thought, I am going to look at this over the next few days. Please coordinate with me before commiting to the branch. Thanks again! 🙏

**Review by YakDriver — APPROVED** [review]
Looks great! 🎉 

Acceptance tests in commercial (`us-west-2`) (Amazon MQ is not supported in GovCloud):
```
--- PASS: TestAccAWSMqBroker_rabbitmq (540.89s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1127.84s)
--- PASS: TestAccAWSMqBroker_disappears (1130.68s)
--- PASS: TestAccAWSMqBroker_updateTags (1160.50s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1161.53s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1163.69s)
--- PASS: TestAccAWSMqBroker_basic (1188.40s)
--- PASS: TestAccAWSMqBroker_updateSecurityGroup (1454.75s)
--- PASS: TestAccAWSMqBroker_updateUsers (1506.68s)
--- PASS: TestAccAWSMqBroker_allFieldsDefaultVpc (1717.77s)
--- PASS: TestAccAWSMqBroker_allFieldsCustomVpc (1755.67s)
```

**Comment by YakDriver** [comment]
This PR also adds enumeration validations per #14601

**Pull request merged**
